### PR TITLE
feat(split): add notify_stop_split rpc

### PR DESCRIPTION
**Pull request by hycdong**
This pull request is the last pr about pause and cancel split (previous pr: #679, #681). When all replicas in group pause or cancel split succeed, primary will send notify_stop_split_request to meta server. If type is pause, meta server will set split_status from PAUSING into PAUSED. If type is cancel, meta server will set this partition not_split, if all partitions of this table cancel split succeed, meta server will update partition_count into original partition_count on zk.

### CODE REVIEW

File: src/meta/meta_split_service.cpp [hunk 1]
```diff
@@ -429,5 +429,88 @@ void meta_split_service::do_control_all(std::shared_ptr<app_state> app, control_
...
+    dassert_f(app != nullptr, "app({}) is not existed", request.app_name);
+    dassert_f(app->is_stateful, "app({}) is stateless currently", request.app_name);
+    dassert_f(request.meta_split_status == split_status::PAUSING ||
+                  request.meta_split_status == split_status::CANCELING,
+              "invalid split_status({})",
+              dsn::enum_to_string(request.meta_split_status));
+
+    const std::string stop_type =
+        rpc.request().meta_split_status == split_status::PAUSING ? "pause" : "cancel";
+    auto iter = app->helpers->split_states.status.find(request.parent_gpid.get_partition_index());
```
Comment (levy5307) [inline src/meta/meta_split_service.cpp:447]: const auto iter

File: src/meta/meta_split_service.cpp [hunk 2]
```diff
@@ -429,5 +429,88 @@ void meta_split_service::do_control_all(std::shared_ptr<app_state> app, control_
...
+    zauto_write_lock(app_lock());
+    std::shared_ptr<app_state> app = _state->get_app(request.app_name);
+    dassert_f(app != nullptr, "app({}) is not existed", request.app_name);
+    dassert_f(app->is_stateful, "app({}) is stateless currently", request.app_name);
+    dassert_f(request.meta_split_status == split_status::PAUSING ||
+                  request.meta_split_status == split_status::CANCELING,
+              "invalid split_status({})",
+              dsn::enum_to_string(request.meta_split_status));
+
+    const std::string stop_type =
```
Comment (levy5307) [inline src/meta/meta_split_service.cpp:445]: const std::string &stop_type

File: src/meta/meta_split_service.cpp [hunk 3]
```diff
@@ -429,5 +429,88 @@ void meta_split_service::do_control_all(std::shared_ptr<app_state> app, control_
...
+void meta_split_service::do_cancel_partition_split(std::shared_ptr<app_state> app,
+                                                   notify_stop_split_rpc rpc)
+{
+    auto on_write_storage_complete = [app, rpc, this]() {
+        ddebug_f("app({}) update partition count on remote storage, new partition count is {}",
+                 app->app_name,
+                 app->partition_count / 2);
+        zauto_write_lock l(app_lock());
+        app->partition_count /= 2;
+        for (int i = app->partition_count; i < app->partition_count * 2; ++i) {
```
Comment (levy5307) [inline src/meta/meta_split_service.cpp:502]:
```
app->partitions.resize(app->partition_count);
app->helpers->contexts.resize(app->partition_count);
```

Comment (levy5307) [inline src/meta/meta_split_service.cpp:502]:
```
app->partitions.resize(app->partition_count);
app->helpers->contexts.resize(app->partition_count);
```

**Review by neverchanje — APPROVED** [review]

**Review by acelyc111 — APPROVED** [review]

**Pull request merged**